aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3FileIOIntegration.java

```diff
@@ -0,0 +1,226 @@
+    ListAliasesResponse listAliasesResponse = kms.listAliases(
+        ListAliasesRequest.builder().keyId(response.ssekmsKeyId()).build());
+    Assert.assertTrue(listAliasesResponse.hasAliases());
+    Assert.assertEquals(1, listAliasesResponse.aliases().size());
+    Assert.assertEquals("alias/aws/s3", listAliasesResponse.aliases().get(0).aliasName());
+  }
+
+  @Test
+  public void testServerSideCustomEncryption() throws Exception {
+    // generate key
+    KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
+    keyGenerator.init(256, new SecureRandom());
+    SecretKey secretKey = keyGenerator.generateKey();
+    Base64.Encoder encoder = Base64.getEncoder();
+    String encodedKey = new String(encoder.encode(secretKey.getEncoded()), StandardCharsets.UTF_8);
+    // generate md5
+    MessageDigest digest = MessageDigest.getInstance("MD5");
+    String md5 = new String(encoder.encode(digest.digest(secretKey.getEncoded())), StandardCharsets.UTF_8);
+
+    AwsProperties properties = new AwsProperties();
+    properties.setS3FileIoSseType(AwsProperties.S3FILEIO_SSE_TYPE_CUSTOM);
+    properties.setS3FileIoSseKey(encodedKey);
+    properties.setS3FileIoSseMd5(md5);
+    S3FileIO s3FileIO = new S3FileIO(clientFactory::s3, properties);
+    write(s3FileIO);
+    validateRead(s3FileIO);
+    GetObjectResponse response = s3.getObject(
+        GetObjectRequest.builder().bucket(bucketName).key(objectKey)
+            .sseCustomerAlgorithm(ServerSideEncryption.AES256.name())
+            .sseCustomerKey(encodedKey)
+            .sseCustomerKeyMD5(md5)
+            .build()).response();
+    Assert.assertNull(response.serverSideEncryption());
+    Assert.assertEquals(ServerSideEncryption.AES256.name(), response.sseCustomerAlgorithm());
+    Assert.assertEquals(md5, response.sseCustomerKeyMD5());
+  }
+
+  @Test
+  public void testACL() throws Exception {
+    AwsProperties properties = new AwsProperties();
+    properties.setS3FileIoAcl(ObjectCannedACL.BUCKET_OWNER_FULL_CONTROL);
+    S3FileIO s3FileIO = new S3FileIO(clientFactory::s3, properties);
+    write(s3FileIO);
+    validateRead(s3FileIO);
+    GetObjectAclResponse response = s3.getObjectAcl(
+        GetObjectAclRequest.builder().bucket(bucketName).key(objectKey).build());
+    Assert.assertTrue(response.hasGrants());
+    Assert.assertEquals(1, response.grants().size());
+    Assert.assertEquals(Permission.FULL_CONTROL, response.grants().get(0).permission());
+  }
+
+  @Test
+  public void testClientFactorySerialization() throws Exception {
+    S3FileIO fileIO = new S3FileIO(clientFactory::s3);
+    write(fileIO);
+    byte [] data = SerializationUtils.serialize(fileIO);
+    S3FileIO fileIO2 = SerializationUtils.deserialize(data);
+    validateRead(fileIO2);
+  }
+
+  private void write(S3FileIO s3FileIO) throws Exception {
+    OutputFile outputFile = s3FileIO.newOutputFile(objectUri);
+    OutputStream outputStream = outputFile.create();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *outputStream*. The resource is referenced at line 214. The resource is closed at line 215. There are other execution paths that do not contain closure statements, for example, when IoUtils.copy throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).